File: tasks/circuit-breaker/typescript/environment/src/parse.ts
export type State = "closed" | "open" | "half-open";
export type Outcome = "success" | "failure" | "neutral";
export type Recorded = Outcome | "rejected";

export type CallOutcome =
  | { kind: "ok" }
  | { kind: "error" }
  | { kind: "status"; status: number };

export interface Call {
  at: number;
  target: string;
  outcome: CallOutcome;
}

export interface Config {
  threshold: number;
  cooldownMs: number;
  halfOpenLimit: number;
  failureStatuses: number[];
}

export interface Document {
  config: Config;
  calls: Call[];
}

const TOP_LEVEL = ["calls", "config"];

/**
 * Accept the parsed input as a run description. The current implementation
 * checks only the top-level key set and trusts every value below it.
 */
export function parseDocument(value: unknown): Document {
  const keys = Object.keys(value ?? {})
    .sort()
    .join(",");
  if (keys !== TOP_LEVEL.join(",")) {
    throw new Error("malformed document");
  }
  return value as Document;
}

export function parseConfig(value: Config): Config {
  return {
    threshold: value.threshold,
    cooldownMs: value.cooldownMs,
    halfOpenLimit: value.halfOpenLimit,
    failureStatuses: value.failureStatuses,
  };
}

/** Every outcome that is not a success counts as a failure. */
export function classify(
  outcome: CallOutcome,
  failureStatuses: number[],
): Outcome {
  return outcome.kind === "ok" ? "success" : "failure";
}

File: tasks/circuit-breaker/python-typed/environment/src/parse.py
"""Read the configuration and classify call outcomes."""

from typing import Any, Literal, TypedDict, cast

State = Literal["closed", "open", "half-open"]
Outcome = Literal["success", "failure", "neutral"]
Recorded = Literal["success", "failure", "neutral", "rejected"]

TOP_LEVEL = {"config", "calls"}


class CallOutcome(TypedDict, total=False):
    kind: str
    status: int


class Call(TypedDict):
    at: int
    target: str
    outcome: CallOutcome


class Config(TypedDict):
    threshold: int
    cooldownMs: int
    halfOpenLimit: int
    failureStatuses: list[int]


class Document(TypedDict):
    config: Config
    calls: list[Call]


def parse_document(value: object) -> Document:
    """Accept the parsed input as a run description.

    The current implementation checks only the top-level key set and trusts
    every value below it.
    """
    if not isinstance(value, dict) or set(value) != TOP_LEVEL:
        raise ValueError("malformed document")
    return cast(Document, value)


def parse_config(value: Config) -> Config:
    return {
        "threshold": value["threshold"],
        "cooldownMs": value["cooldownMs"],
        "halfOpenLimit": value["halfOpenLimit"],
        "failureStatuses": value["failureStatuses"],
    }


def classify(outcome: CallOutcome, failure_statuses: list[int]) -> Outcome:
    """Every outcome that is not a success counts as a failure."""
    return "success" if outcome["kind"] == "ok" else "failure"
